drivers/adc/adc_npcx.c

```diff
@@ -0,0 +1,374 @@
+
+	LOG_DBG("Start ADC scan conversion and ADCCNF,ADCCS are (%04X,%04X)\n",
+			inst->ADCCNF, inst->ADCCS);
+}
+
+static int adc_npcx_start_read(const struct device *dev,
+					const struct adc_sequence *sequence)
+{
+	struct adc_npcx_data *const data = DRV_DATA(dev);
+	int error = 0;
+
+	if (!sequence->channels ||
+	    (sequence->channels & ~BIT_MASK(NPCX_ADC_CH_COUNT))) {
+		LOG_ERR("Invalid ADC channels");
+		return -EINVAL;
+	}
+
+	/* Fixed 10 bit resolution of npcx ADC */
+	if (sequence->resolution != 10) {
+		LOG_ERR("Unfixed 10 bit ADC resolution");
+		return -ENOTSUP;
+	}
+
+	error = adc_npcx_validate_buffer_size(dev, sequence);
+	if (error) {
+		LOG_ERR("ADC buffer size too small");
+		return error;
+	}
+
+	/* Save ADC sequence sampling buffer */
+	data->buffer = sequence->buffer;
+
+	/* Start ADC conversion */
+	adc_context_start_read(&data->ctx, sequence);
+	error = adc_context_wait_for_completion(&data->ctx);
+
+	return error;
+}
+
+/* ADC api functions */
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_npcx_data *const data =
+		CONTAINER_OF(ctx, struct adc_npcx_data, ctx);
+
+	data->repeat_buffer = data->buffer;
+	data->channels = ctx->sequence.channels;
+
+	/* Start ADC scan conversion */
+	adc_npcx_start_scan(data->adc_dev);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_npcx_data *const data =
+		CONTAINER_OF(ctx, struct adc_npcx_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static int adc_npcx_channel_setup(const struct device *dev,
+				 const struct adc_channel_cfg *channel_cfg)
+{
+	const struct adc_npcx_config *const config = DRV_CONFIG(dev);
+	uint8_t channel_id = channel_cfg->channel_id;
+
+	if (channel_id >= NPCX_ADC_CH_COUNT) {
+		LOG_ERR("Invalid channel %d", channel_id);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
+		LOG_ERR("Unsupported channel acquisition time");
+		return -ENOTSUP;
+	}
+
+	if (channel_cfg->differential) {
+		LOG_ERR("Differential channels are not supported");
+		return -ENOTSUP;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("Unsupported channel gain %d", channel_cfg->gain);
+		return -ENOTSUP;
+	}
+
+	if (channel_cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("Unsupported channel reference");
+		return -ENOTSUP;
+	}
+
+	/* Configure pin-mux for ADC channel */
+	soc_pinctrl_mux_configure(config->alts_list + channel_cfg->channel_id,
+			1, 1);
+	LOG_DBG("ADC channel %d, alts(%d,%d)", channel_cfg->channel_id,
+			config->alts_list[channel_cfg->channel_id].group,
+			config->alts_list[channel_cfg->channel_id].bit);
+
+	return 0;
+}
+
+static int adc_npcx_read(const struct device *dev,
+			const struct adc_sequence *sequence)
+{
+	struct adc_npcx_data *const data = DRV_DATA(dev);
+	int error;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	error = adc_npcx_start_read(dev, sequence);
```

Comment:
Should this `adc_npcx_read` function call `adc_context_wait_for_completion` instead of having `adc_npcx_start_read` call the function?

Otherwise the read_async below and this synchronous read do the same thing.

Reply:
adc_npcx_start_read() doesn't determine the reading is sync or async. Per my understanding, it depends on the third parameter of adc_context_lock() and adc_context_unlock().

Reply:
I see that now. The API seems a little weird, but that isn't code you wrote. This LGTM now.